rfd/XXXX-cloud-agent-upgrades.md

```diff
@@ -0,0 +1,318 @@
+the rate at which discovery occurs should be able to be temporarily increased and/or triggered by poor teleport
+health (esp. crashes). This also means that teleport can try to "force" an update if it has been unhealthy for too
+long by exiting.
+
+For the initial block of cloud-stable work, two variants of the upgrade process will exist. When running k8s, both discovery
+and restart will be handled by a separate controller. When running on a VM, teleport discovery will be handled indipendently
+of teleport by a periodic operation that pulls from the appropriate `apt`/`yum` repo, and restart will take the
+form of a graceful exit initiated by the agent itself.
+
+For all variants, the agent will be required to gather maintenance window information from the auth server. In the interest of maximum
+simplicity/extensibility, we won't be mirroring the current maintenance window schema as it is represented in the teleport cloud database.
+Instead, we will select a generic representation that preserves flexibility:
+
+```proto3
+message AgentMaintenanceWindowsRequest {
+    string TeleportVersion = 1;
+}
+
+message AgentMaintenanceWindows {
+    repeated AgentMaintenanceWindow Windows = 1;
+    bool AppearsOutdatedHint = 2;
+}
+
+message AgentMaintenanceWindow {
+  google.protobuf.Timestamp StartTime = 1;
+  google.protobuf.Timestamp EndTime = 2;
+}
+
+service AuthServer {
+    // ...
+
+    rcp GetAgentMaintenanceWindows(AgentMaintenanceWindowsRequest) returns (AgentMaintenanceWindowsResponse);
+}
+
+// note: opting for pull-based over push-based since agents will want to discover the window immediately on
+// startup under some conditions, but do *not* need to rediscover the window upon auth connectivity interruption.
+```
+
+The auth server will be permitted to return as many or as few maintenance windows as it would like. Agents will filter
+down the provided windows to those whose `EndTime` is in the future, and whose `StartTime` is before `now + max_maintenance_interval`
+(likely 1.5-2 weeks). If at least one upgrade window exists post-filter, at least one attempt at discovery + restart aught
+to be made within one of the provided windows. Additionally, at least one discovery + restart attempt must be made at least
+every `max_maintenance_interval` regardless.
+
+The additional `AppearsOutdatedHint` flag will permit the auth server to suggest that the *specific* client version sent
+in the request is too old. By default, agents should ignore windows that started *before* the start time of
+the current process, and should endeavor to spread out restarts through semi-random window selection. The presence of
+this flag indicates that these behaviors should be overridden (likely due to a critical patch becoming available).
+
+Because the control plane might also restart during a maintenance window, direct contact with the auth server should
```

Comment:
Being able to upgrade without being able to contact the control plane was one of the design guidelines for this proposal.  We wanted very old teleport agents and/or very bugged teleport agents to still be able to be upgraded, and ideally at a fairly fast cadence.

I definitely see the potential concern of having a "thundering herd" type issue... if new agents induce errors or excess load on the control plane, then that can result in the bad agent version spreading even faster.  That being said, I don't see a good way to prevent that without also by necessity causing bugged agents to become orphaned permanently.  Because teleport is sometimes the _only_ means of accessing a server, I think we have to treat permanent loss of access as a strictly worse outcome than any temporary/recoverable outage.